sys/net/application_layer/gcoap/dns.c

```diff
@@ -0,0 +1,683 @@
+
+        coap_block_object_init(&block, 0, CONFIG_GCOAP_DNS_BLOCK_SIZE, 0);
+        return coap_opt_add_block2_control(pdu, &block);
+    }
+    return 0;
+}
+
+static int _add_proxy_uri_opt(coap_pkt_t *pdu, const char *proxy_uri)
+{
+    if (_is_proxied()) {
+        return coap_opt_add_proxy_uri(pdu, proxy_uri);
+    }
+    return 0;
+}
+
+static int _add_req_tag_opt(coap_pkt_t *pdu, _req_ctx_t *context)
+{
+    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
+        return coap_opt_add_opaque(pdu, 292, (uint8_t *)&context->req_tag,
+                                   sizeof(context->req_tag));
+    }
+    return 0;
+}
+
+static int _add_remaining_options(coap_pkt_t *pdu, const char *proxy_uri, _req_ctx_t *context)
+{
+    if (_add_proxy_uri_opt(pdu, proxy_uri) < 0) {
+        DEBUG("gcoap_dns: unable to add Proxy-URI option to request\n");
+        return -ENOBUFS;
+    }
+    /* add request tag to distinguish multiple blockwise requests on-the-air */
+    if (_add_req_tag_opt(pdu, context) < 0) {
+        DEBUG("gcoap_dns: unable to add Request-Tag option to request");
+        return -ENOBUFS;
+    }
+    return 0;
+}
+
+static size_t _dns_msg_compose(void *dns_buf, const char *domain_name,
+                               int family)
+{
+    return dns_msg_compose_query(dns_buf, domain_name, 0, family);
+}
+
+static int _set_remote(const uri_parser_result_t *uri_comp,
+                       sock_udp_ep_t *remote) {
+    assert(uri_comp->host != NULL);
+    memset(remote, 0, sizeof(*remote));
+
+    if (uri_comp->ipv6addr != NULL) {
+        if (ipv6_addr_from_buf((ipv6_addr_t *)remote->addr.ipv6,
+                               uri_comp->ipv6addr,
+                               uri_comp->ipv6addr_len) == NULL) {
+            DEBUG("gcoap_dns: unable to parse IPv6 address %*.s\n",
+                  uri_comp->ipv6addr_len, uri_comp->ipv6addr);
+            return -EINVAL;
+        }
+        remote->family = AF_INET6;
+        if (uri_comp->zoneid != NULL) {
+            netif_t *netif = netif_get_by_name_buffer(uri_comp->zoneid,
+                                                      uri_comp->zoneid_len);
+            if (netif == NULL) {
+                DEBUG("gcoap_dns: unable to interpret zoneid %*.s\n",
+                      uri_comp->zoneid_len, uri_comp->zoneid);
+                return -ENOENT;
+            }
+            remote->netif = netif_get_id(netif);
+        }
+        else {
+            remote->netif = SOCK_ADDR_ANY_NETIF;
+        }
+    }
+    else if (ipv4_addr_from_buf((ipv4_addr_t *)remote->addr.ipv4,
+                                uri_comp->host, uri_comp->host_len) != NULL) {
+        remote->family = AF_INET;
+    }
+    else if (IS_USED(MODULE_SOCK_DNS)) {
+        char hostname[uri_comp->host_len + 1];
+
+        strncpy(hostname, uri_comp->host, uri_comp->host_len);
+        hostname[uri_comp->host_len] = '\0';
+        int res = sock_dns_query(hostname, remote->addr.ipv6, AF_UNSPEC);
```

Comment:
(Summarizing H&A discussion) I'm OK with having it in here, but will suggest a comment for the servr setting docs.

Reply:
This is purely optional (if the `sock_dns` module is not available, it won't as with any hostname in other applications, be compiled in). Since we _are_ handling valid CoAP-URIs here, we _should_ be able to handle it. If the user decides they don't need that, `sock_dns` won't be included, so name resolution for the DoC server URI won't be provided.